operator/pkg/controller/virtualippool/controller.go

```diff
@@ -0,0 +1,425 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	metalk8sscalitycomv1alpha1 "github.com/scality/metalk8s/operator/api/v1alpha1"
+	"github.com/scality/metalk8s/operator/pkg/controller/utils"
+)
+
+type VirtualIPPoolReconciler struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+
+	instance metalk8sscalitycomv1alpha1.VirtualIPPool
+	// Some cache to ensure we don't reuse Virtual Router IDs, even for different pools
+	usedVRID       map[int]bool
+	nodeList       corev1.NodeList
+	configChecksum string
+}
+
+const (
+	// Name of the controller
+	controllerName = "virtualippool-controller"
+
+	// Name of the component
+	componentName = "metalk8s-vips"
+
+	// Name of the application
+	appName = "virtualippool"
+
+	// Annotation key to store Config checksum
+	annotationChecksumName = "checksum/config"
+
+	// ConfigMap key for HL config
+	hlConfigMapKey = "hl-config.yaml"
+)
+
+var log = logf.Log.WithName(controllerName)
+
+// Create a new VirtualIPPoolReconciler
+func newVirtualIPPoolReconciler(mgr ctrl.Manager) *VirtualIPPoolReconciler {
+	return &VirtualIPPoolReconciler{
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+}
+
+// Add create the new Reconciler
+func Add(mgr ctrl.Manager) error {
+	reconciler := newVirtualIPPoolReconciler(mgr)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metalk8sscalitycomv1alpha1.VirtualIPPool{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&appsv1.DaemonSet{}).
+		Complete(reconciler)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *VirtualIPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := log.WithValues("Request.Name", req.Name).WithValues("Request.Namespace", req.Namespace)
+	reqLogger.Info("reconciling VirtualIPPool: START")
+	defer reqLogger.Info("reconciling VirtualIPPool: STOP")
+
+	err := r.client.Get(ctx, req.NamespacedName, &r.instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Nothing to do, all objects that should be deleted are
+			// automatically grabage collected because of owner reference
+			return utils.EndReconciliation()
+		}
+		return utils.Requeue(err)
+	}
+
+	// Retrieve all nodes
+	if err := r.client.List(ctx, &r.nodeList); err != nil {
+		return utils.Requeue(err)
+	}
+
+	// Retrieve all configured Virtual Router IDs
+	r.usedVRID = make(map[int]bool, 255)
+	pools := metalk8sscalitycomv1alpha1.VirtualIPPoolList{}
+	if err := r.client.List(ctx, &pools); err != nil {
+		return utils.Requeue(err)
+	}
+	for _, pool := range pools.Items {
+		r.retrieveVRID(ctx, pool)
```

Comment:
Shall we handle the possible error here ?

Reply:
If I understand correctly the sequence here `r.retrieveVRID` is mutating `r.usedVRID` so when we use it we have to be aware of the exact sequence we need to follow. IMO we should try to avoid this imperative paradigm. Maybe we can use an intermediate facade 🤔

Reply:
> Shall we handle the possible error here ?

Indeed, I forgot about it, I will add it

Reply:
I agree it's a bit weird, but I do not see any "clean" solution to do it

> If I understand correctly the sequence here `r.retrieveVRID` is mutating `r.usedVRID`

Yes

> so when we use it we have to be aware of the exact sequence we need to follow

WDYM ?

To explain a bit the behavior here, we retrieve all the defined pools, we iter on those pools to retrieve the "potential" already defined VRID to "cache" them, so that if we need to pick a new VRID we pick one that is not already used

> Maybe we can use an intermediate facade

I'm not sure to understand how it could work, except if we have "something" that basically manages all the VirtualIPPools at once, but it makes things complicated and not really good as well IMHO.

Reply:
> Shall we handle the possible error here ?

Done